server/ec2spotmanager/tasks.py

```diff
@@ -277,230 +210,415 @@ def _start_pool_instances(pool, config, count=1):
-                    instance_ids_by_region[region].append(successful_requests[req_id]['instance_id'])
-
-                    instances_created = True
-
-                for req_id in failed_requests:
-                    instance = instances_by_ids[req_id]
-                    if failed_requests[req_id]['action'] == 'blacklist':
-                        # request was not fulfilled for some reason.. blacklist this type/zone for a while
-                        inst = instances_by_ids[req_id]
-                        key = "%s:blacklist:%s:%s" % (cloud_provider.get_name(), inst.zone,
-                                                      failed_requests[req_id]['instance_type'])
-                        cache.set(key, "", ex=12 * 3600)
-                        logger.warning("Blacklisted %s for 12h", key)
-                        inst.delete()
-                    elif failed_requests[req_id]['action'] == 'disable_pool':
-                        _update_pool_status(pool, 'unclassifed', 'request failed')
-
-            cloud_instances = cloud_provider.check_instances_state(pool.pk, region)
-
-            for cloud_instance in cloud_instances:
-                debug_cloud_instances_ids_seen.add(cloud_instance)
-
-                if (SPOTMGR_TAG + "-Updatable" not in cloud_instances[cloud_instance]['tags'] or
-                        int(cloud_instances[cloud_instance]['tags'][SPOTMGR_TAG + "-Updatable"]) <= 0):
-                    # The instance is not marked as updatable. We must not touch it because
-                    # a spawning thread is still managing this instance. However, we must also
-                    # remove this instance from the instances_left list if it's already in our
-                    # database, because otherwise our code here would delete it from the database.
-                    if cloud_instance in instance_ids_by_region[region]:
-                        instances_left.remove(instances_by_ids[cloud_instance])
+    logger.debug("-> update_instances(%r, %r)", provider, region)
+
+    try:
+        cloud_provider = CloudProvider.get_instance(provider)
+
+        debug_cloud_instance_ids_seen = set()
+        debug_not_updatable_continue = set()
+        debug_not_in_region = {}
+
+        instances = {}  # provider_id -> Instance
+        instances_left = set()  # Instance
+        for instance in Instance.objects.filter(provider=provider, region=region) \
+                .exclude(status_code=INSTANCE_STATE['requested']):
+            instances[instance.instance_id] = instance
+            instances_left.add(instance)
+
+        # reconcile instance state from provider with DB
+        cloud_instances = cloud_provider.check_instances_state(None, region)
+        for cloud_id, cloud_data in cloud_instances.items():
+            debug_cloud_instance_ids_seen.add(cloud_id)
+
+            if (SPOTMGR_TAG + "-Updatable" not in cloud_data['tags'] or
```

Comment:
If I understand your new model correctly, then the whole "Updatable" tag is obsolete now? Earlier, we could have one thread spawning instances while a second thread was updating instances, so this could happen in parallel. This is also one of the most fragile parts of the code from what I remember. We used to have races here that should no longer be possible now. Or am I misunderstanding the new model?

Maybe we should add some kind of logging to confirm that this case is really dead?

Reply:
You are right. I tried to write a test to cover this area, and couldn't think of a scenario where this could happen anymore. I can add a standout log entry and if we don't see anything in the next month or two, remove this block.

Reply:
Yes, that would be perfect :)